File: examples/python/testing-patterns/complex-workflows/expected-failures-workflow.py
@durable_step
def validate_input(step_context: StepContext, value: int) -> int:
    if value < 0:
        raise ValueError("Value must be non-negative")
    return value

@durable_execution
def handler(event: dict, context: DurableContext) -> int:
    value = event.get("value", 0)
    validated = context.step(validate_input(value), name="validate")
    return validated

File: examples/java/testing-patterns/complex-workflows/expected-failures-workflow.java
// Expected failures workflow handler
public class ValidationHandler extends DurableHandler<Map, Integer> {
    @Override
    public Integer handleRequest(Map input, DurableContext ctx) {
        int value = (int) input.getOrDefault("value", 0);

        int validated = ctx.step("validate", Integer.class, stepCtx -> {
            if (value < 0) {
                throw new IllegalArgumentException("Value must be non-negative");
            }
            return value;
        });

        return validated;
    }
}
